File: boj/11729.cpp
#include <iostream>
#include <cmath>

using namespace std;

void solve(int n, int start, int end) {
    if (n == 0) return;
    solve(n - 1, start, 6 - start - end);
    cout << start << " " << end << "\n";
    solve(n - 1, 6 - start - end, end);
}

int main(void) {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL); cout.tie(NULL);
    int n; cin >> n;
    cout << (int)pow(2, n) - 1 << "\n";
    solve(n, 1, 3);
    return 0;
}

File: boj/11729.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class Main {

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        StringBuilder sb = new StringBuilder();
        sb.append((int)Math.pow(2, n) - 1).append("\n");
        solve(n, 1, 3, sb);
        System.out.println(sb);
    }

    public static void solve(int n, int start, int end, StringBuilder sb) {
        if(n == 0) return;
        solve(n - 1, start, 6 - start - end, sb);
        sb.append(start).append(" ").append(end).append("\n");
        solve(n - 1, 6 - start - end, end, sb);
    }
}